## Importing Libraries

In [2]:
# importing importi=ant libraries
import pandas as pd
import numpy as np 
from matplotlib import pyplot as plt 
import seaborn as sns
import re

## Loading Dataset

In [3]:
# loading Dataset in notebook
df = pd.read_csv("clash_data.csv")

## Data Preview

In [4]:
df.sample(5)

,Card,Card Level (Spawn Level),Cost,Count,Crown Tower Damage,Damage,Damage per second,Death Damage,Health (+Shield),Hit Speed,...,Maximum Spawned,Radius,Range,Spawn DPS,Spawn Damage,Spawn Health,Spawn Speed,Spawner Health,Troop Spawned,Type
26,Ice Golem,NaN,2.0,1,NaN,70,28,70.0,"1,047",2.5,...,NaN,NaN,1,NaN,NaN,NaN,NaN,NaN,NaN,Troops and Defenses
1,Baby Dragon,NaN,4.0,1,NaN,133,83,0.0,"1,064",1.6,...,NaN,NaN,3.5,NaN,NaN,NaN,NaN,NaN,NaN,Troops and Defenses
43,P.E.K.K.A.,NaN,7.0,1,NaN,678,376,0.0,"3,458",1.8,...,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN,NaN,Troops and Defenses
21,Goblins,NaN,2.0,3,NaN,106,96,0.0,169,1.1,...,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN,NaN,Troops and Defenses
32,Lava Hound,NaN,7.0,1,NaN,45,34,0.0,"3,000",1.3,...,NaN,NaN,2,NaN,NaN,NaN,NaN,NaN,NaN,Troops and Defenses


In [5]:
df.shape

(70, 21)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Card                      70 non-null     object 
 1   Card Level (Spawn Level)  6 non-null      object 
 2   Cost                      68 non-null     float64
 3   Count                     57 non-null     object 
 4   Crown Tower Damage        7 non-null      float64
 5   Damage                    64 non-null     object 
 6   Damage per second         57 non-null     object 
 7   Death Damage              57 non-null     float64
 8   Health (+Shield)          57 non-null     object 
 9   Hit Speed                 57 non-null     object 
 10  Level                     64 non-null     float64
 11  Maximum Spawned           6 non-null      float64
 12  Radius                    7 non-null      float64
 13  Range                     57 non-null     object 
 14  Spawn DPS   

- also Datatypes of some important numeric stats are object type so we need to change that also. 

In [7]:
df.isnull().sum()

Card                         0
Card Level (Spawn Level)    64
Cost                         2
Count                       13
Crown Tower Damage          63
Damage                       6
Damage per second           13
Death Damage                13
Health (+Shield)            13
Hit Speed                   13
Level                        6
Maximum Spawned             64
Radius                      63
Range                       13
Spawn DPS                   64
Spawn Damage                64
Spawn Health                64
Spawn Speed                 64
Spawner Health              64
Troop Spawned               64
Type                         0
dtype: int64

In [8]:
df.columns

Index(['Card', 'Card Level (Spawn Level)', 'Cost', 'Count',
       'Crown Tower Damage', 'Damage', 'Damage per second', 'Death Damage',
       'Health (+Shield)', 'Hit Speed', 'Level', 'Maximum Spawned', 'Radius',
       'Range', 'Spawn DPS', 'Spawn Damage', 'Spawn Health', 'Spawn Speed',
       'Spawner Health', 'Troop Spawned', 'Type'],
      dtype='object')

- here some columns names are looking differnt from others like Card Level (Spawn Level) and Health (+Shield)
- As a player of game i know their might be values in such format so need to check values format and derive immportant data.

In [9]:
df['Health (+Shield)'].unique()

array(['254', '1,064', '1,396', '780', '636', '67', '756', '1,672', '311',
       '1,596', '742', '696 (+696)', '977 (+266)', '216', '590', '970',
       '1,010', '91', '3,344', '2,660', '169/110', '169', '4,256', '864',
       '86 (+199)', '1,408', '1,047', '190', '665', '1,070', '1,399',
       '3,000', '179', '990', '695', '1,056', '1,000', '1,272', '598',
       '750', '3,458', '1,463', '2,544', '1,200', '110', '954', '1,548',
       '696', '1,330', nan], dtype=object)

In [10]:
df['Card Level (Spawn Level)'].unique()

array([nan, '7 (9)', '4 (9)', '1 (9)'], dtype=object)

- as we thought so their are multiple data in one column so we need to derive them seperately for later stages of Analysis
- also their are values which might cause problem in type change later we need to handle them as well like 1,047, 169/110

=> also names of columns are very long so we will change them for better handling 

## Handling DataSet

In [11]:
# Renaming the columns for better code writting and speed
df.rename(columns={
    "Card":"card",
    "Cost":"cost",
    "Count":"count",
    "Damage":"damage",
    "Death Damage":"death_damage",
    "Health (+Shield)":"sheild",
    "Card Level (Spawn Level)": "card_level",
    "Crown Tower Damage": "tower_damage",
    "Damage per second": "dps",
    "Hit Speed":"hit_speed",
    "Level":"level",
    "Maximum Spawned":"max_spawn",
    "Radius":"radius",
    "Range":"range",
    "Spawn Damage":"spawn_damage",
    "Spawn DPS":"spawn_dps",
    "Spawn Health":"spawn_health",
    "Spawn Speed":"spawn_speed",
    "Spawner Health":"spawner_health",
    "Spawn Damage":"spawn_damage",
    "Troop Spawned":"troop_spawned",
    "Type":"type"
}, inplace=True)

In [12]:
df["damage"].unique()

array(['86', '133', '798', '160', '159', '67', '246', '176', '271', '239',
       '127', '203', '206', '93', '200', '254', '280', '169', '211',
       '172', '106/50', '106', '259', '53', '264', '70', '95', '69',
       '30-350', '35-704', '45', '258', '598', '84', '228', '260', '678',
       '325', '140', '1,300', '50', '135', '26', nan, '243', '572', '864',
       '600', '1,232', '240'], dtype=object)

In [13]:
df["dps"].unique()

array(['71', '83', '266', '160', '106', '67', '0', '110', '142', '95',
       '158', '169', '147', '132', '111', '116', '140', '114', '96/38',
       '96', '103', '21', '176', '28', '40', '75-875', '87-1,760', '144',
       '34', '45', '285', '172', '332', '133', '84', '173', '376', '216',
       '46', '93', '260', '38', '168', '98', '163', '104', nan],
      dtype=object)

In [14]:
df["hit_speed"].unique()

array(['1.2', '1.6', '3', '1', '1.5', '0', '1.9', '2.5', '0.8', '0.7',
       '1.8', '2.4', '1.1/1.3', '1.1', '1.7', '0.4', '1.3', '5', '1.4',
       '0.25', nan], dtype=object)

- Their are some problematic values in all these columns so we need to clean them first before typechange 

In [15]:
def clean_damage(val):
    if pd.isnull(val):
        return np.nan
    
    val = str(val).replace(',', '').strip()
    
    # Case 1: Range (e.g., 30-350)
    if '-' in val:
        parts = re.split(r'-', val)
        if len(parts) == 2:
            return (float(parts[0]) + float(parts[1])) / 2
    
    # Case 2: Fraction (e.g., 106/50)
    if '/' in val:
        parts = re.split(r'/', val)
        if len(parts) == 2:
            return float(parts[0]) / float(parts[1])
    
    # Case 3: Normal number
    try:
        return float(val)
    except:
        return np.nan

In [16]:
df['damage'] = df['damage'].apply(clean_damage)
df['dps'] = df['dps'].apply(clean_damage)
df['hit_speed'] = df['hit_speed'].apply(clean_damage)
df["range"]=df["range"].apply(clean_damage)

In [17]:
# Extracted the values of spawn unit level and original card level from card_level. 
df[['card_level_f', 'spawn_level']] = df['card_level'].str.extract(
    r'(\d+)\s*\((\d+)\)'
)

In [18]:
# checnking if changes are done correctly 
df['spawn_level'].unique()

array([nan, '9'], dtype=object)

In [19]:
# dropping the original column
df.drop("card_level",axis = 1, inplace = True)

In [20]:
df.rename(columns = {"card_level_f" : "card_level"}, inplace = True)

In [21]:
df.isnull().sum()

card               0
cost               2
count             13
tower_damage      63
damage             6
dps               13
death_damage      13
sheild            13
hit_speed         13
level              6
max_spawn         64
radius            63
range             13
spawn_dps         64
spawn_damage      64
spawn_health      64
spawn_speed       64
spawner_health    64
troop_spawned     64
type               0
card_level        64
spawn_level       64
dtype: int64

In [22]:
# must handle improper format values first then extract fearture 
df["sheild"].unique()

array(['254', '1,064', '1,396', '780', '636', '67', '756', '1,672', '311',
       '1,596', '742', '696 (+696)', '977 (+266)', '216', '590', '970',
       '1,010', '91', '3,344', '2,660', '169/110', '169', '4,256', '864',
       '86 (+199)', '1,408', '1,047', '190', '665', '1,070', '1,399',
       '3,000', '179', '990', '695', '1,056', '1,000', '1,272', '598',
       '750', '3,458', '1,463', '2,544', '1,200', '110', '954', '1,548',
       '696', '1,330', nan], dtype=object)

In [23]:
def split_hp(val):
    if pd.isnull(val):
        return pd.Series([np.nan, np.nan])
    
    val = str(val).replace(',', '').strip()
    
    # Case 1: Shield format (e.g., 696 (+696))
    match = re.match(r'(\d+)\s*\(\+(\d+)\)', val)
    if match:
        base = float(match.group(1))
        shield = float(match.group(2))
        return pd.Series([base, shield])
    
    # Case 2: Fraction-like (e.g., 169/110) → treat as base/shield
    if '/' in val:
        parts = val.split('/')
        if len(parts) == 2:
            return pd.Series([float(parts[0]), float(parts[1])])
    
    # Case 3: Normal value → only base HP
    try:
        return pd.Series([float(val), 0])  # no shield
    except:
        return pd.Series([np.nan, np.nan])

In [24]:
# values are allocated to shield and base from original column and then it droppep
df[['base_hp', 'shield_hp']] = df['sheild'].apply(split_hp)
df.drop('sheild', axis=1, inplace=True)

## changing datatype 

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   card            70 non-null     object 
 1   cost            68 non-null     float64
 2   count           57 non-null     object 
 3   tower_damage    7 non-null      float64
 4   damage          64 non-null     float64
 5   dps             57 non-null     float64
 6   death_damage    57 non-null     float64
 7   hit_speed       57 non-null     float64
 8   level           64 non-null     float64
 9   max_spawn       6 non-null      float64
 10  radius          7 non-null      float64
 11  range           57 non-null     float64
 12  spawn_dps       6 non-null      float64
 13  spawn_damage    6 non-null      float64
 14  spawn_health    6 non-null      float64
 15  spawn_speed     6 non-null      float64
 16  spawner_health  6 non-null      object 
 17  troop_spawned   6 non-null      objec

In [26]:
# Convert safely (handles NaN automatically)
cols_to_convert = ['count', 'range', 'spawner_health']

for col in cols_to_convert:
    df[col] = pd.to_numeric(df[col], errors='coerce')

In [27]:
numeric_cols = [
    'cost', 'tower_damage', 'damage', 'dps', 'death_damage',
    'hit_speed', 'level', 'max_spawn', 'radius',
    'spawn_dps', 'spawn_damage', 'spawn_health', 'spawn_speed',
    'base_hp', 'shield_hp'
]

df[numeric_cols] = df[numeric_cols].astype(float)

In [85]:
df.isnull().sum()

card                 0
cost                 0
count                0
tower_damage         0
damage               0
dps                  7
death_damage         0
hit_speed           13
level                0
max_spawn           64
radius              63
range               13
spawn_dps           64
spawn_damage        64
spawn_health        64
spawn_speed         64
spawner_health      67
troop_spawned        0
type                 0
card_level          64
spawn_level         64
base_hp             13
shield_hp           13
can_attack_tower     0
dtype: int64

## Handling Null Values 

In [29]:
# replaced null values with none as all cards does not spawn units 
df['troop_spawned'] = df['troop_spawned'].fillna('None')

In [30]:
df['troop_spawned'].unique()

array(['None', 'Barbarians', 'Fire Spirits', 'Goblins', 'Spear Goblins',
       'Skeletons'], dtype=object)

In [34]:
df['cost'] = df['cost'].fillna(df['cost'].median())

In [52]:
df.sample(10)

,card,cost,count,tower_damage,damage,dps,death_damage,hit_speed,level,max_spawn,...,spawn_damage,spawn_health,spawn_speed,spawner_health,troop_spawned,type,card_level,spawn_level,base_hp,shield_hp
52,Three Musketeers,9.0,3.0,NaN,176.0,160.0,0.0,1.1,7.0,NaN,...,NaN,NaN,NaN,NaN,None,Troops and Defenses,NaN,NaN,598.0,0.0
31,Knight,3.0,1.0,NaN,159.0,144.0,0.0,1.1,9.0,NaN,...,NaN,NaN,NaN,NaN,None,Troops and Defenses,NaN,NaN,1399.0,0.0
15,Elite Barbarians,6.0,2.0,NaN,254.0,169.0,0.0,1.5,9.0,NaN,...,NaN,NaN,NaN,NaN,None,Troops and Defenses,NaN,NaN,970.0,0.0
33,Lava Pup,4.0,6.0,NaN,45.0,45.0,0.0,1.0,1.0,NaN,...,NaN,NaN,NaN,NaN,None,Troops and Defenses,NaN,NaN,179.0,0.0
65,Lightning,6.0,0.0,346.0,864.0,NaN,NaN,NaN,4.0,NaN,...,NaN,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN
25,Hog Rider,4.0,1.0,NaN,264.0,176.0,0.0,1.5,7.0,NaN,...,NaN,NaN,NaN,NaN,None,Troops and Defenses,NaN,NaN,1408.0,0.0
41,Musketeer,4.0,1.0,NaN,176.0,160.0,0.0,1.1,7.0,NaN,...,NaN,NaN,NaN,NaN,None,Troops and Defenses,NaN,NaN,598.0,0.0
17,Fire Spirits,2.0,3.0,NaN,169.0,0.0,0.0,0.0,9.0,NaN,...,NaN,NaN,NaN,NaN,None,Troops and Defenses,NaN,NaN,91.0,0.0
35,Mega Minion,3.0,1.0,NaN,258.0,172.0,0.0,1.5,7.0,NaN,...,NaN,NaN,NaN,NaN,None,Troops and Defenses,NaN,NaN,695.0,0.0
3,Bandit,3.0,1.0,NaN,160.0,160.0,0.0,1.0,1.0,NaN,...,NaN,NaN,NaN,NaN,None,Troops and Defenses,NaN,NaN,780.0,0.0


In [50]:
# replaced null with 1 where cards are not spells like witch 1 is deployed
df.loc[df['type'] != 'Damaging Spells', 'count'] = df.loc[df['type'] != 'Damaging Spells', 'count'].fillna(1)

In [51]:
# for spells replaced null with 0 as they are not deployed as units.
df['count'] = df['count'].fillna(0)

In [58]:
# handling tower damaging cards better 
df['tower_damage'].unique()

array([ nan,  98., 229., 346., 240., 493.,  96.,  64.])

In [59]:
# created indicator for later analysis
df['can_attack_tower'] = df['tower_damage'].notnull().astype(int)

In [61]:
# then values are filled with zero 
df['tower_damage'] = df['tower_damage'].fillna(0)

In [63]:
# handling damage 
df['damage'].unique()

array([  86.  ,  133.  ,  798.  ,  160.  ,  159.  ,   67.  ,  246.  ,
        176.  ,  271.  ,  239.  ,  127.  ,  203.  ,  206.  ,   93.  ,
        200.  ,  254.  ,  280.  ,  169.  ,  211.  ,  172.  ,    2.12,
        106.  ,  259.  ,   53.  ,  264.  ,   70.  ,   95.  ,   69.  ,
        190.  ,  369.5 ,   45.  ,  258.  ,  598.  ,   84.  ,  228.  ,
        260.  ,  678.  ,  325.  ,  140.  , 1300.  ,   50.  ,  135.  ,
         26.  ,     nan,  243.  ,  572.  ,  864.  ,  600.  , 1232.  ,
        240.  ])

In [65]:
# as we know only spawner building deal zero damage directly 
df[df['damage'].isnull()]

,card,cost,count,tower_damage,damage,dps,death_damage,hit_speed,level,max_spawn,...,spawn_health,spawn_speed,spawner_health,troop_spawned,type,card_level,spawn_level,base_hp,shield_hp,can_attack_tower
57,Barbarian Hut,7.0,1.0,0.0,NaN,NaN,NaN,NaN,NaN,10.0,...,636.0,14.0,NaN,Barbarians,Spawners,7,9,NaN,NaN,0
58,Furnace,4.0,1.0,0.0,NaN,NaN,NaN,NaN,NaN,12.0,...,91.0,10.0,NaN,Fire Spirits,Spawners,7,9,NaN,NaN,0
59,Goblin Barrel,3.0,1.0,0.0,NaN,NaN,NaN,NaN,NaN,3.0,...,169.0,0.0,0.0,Goblins,Spawners,4,9,NaN,NaN,0
60,Goblin Hut,5.0,1.0,0.0,NaN,NaN,NaN,NaN,NaN,13.0,...,110.0,4.9,NaN,Spear Goblins,Spawners,7,9,NaN,NaN,0
61,Graveyard,5.0,1.0,0.0,NaN,NaN,NaN,NaN,NaN,15.0,...,67.0,0.5,0.0,Skeletons,Spawners,1,9,NaN,NaN,0
62,Tombstone,3.0,1.0,0.0,NaN,NaN,NaN,NaN,NaN,18.0,...,67.0,2.9,422.0,Skeletons,Spawners,7,9,NaN,NaN,0


In [66]:
df['damage'] = df['damage'].fillna(0)

In [69]:
df['dps']

0      71.0
1      83.0
2     266.0
3     160.0
4     106.0
      ...  
65      NaN
66      NaN
67      NaN
68      NaN
69      NaN
Name: dps, Length: 70, dtype: float64

In [70]:
# checking whose values are null helps in such small dataset for better utilization of data 
df[df['dps'].isnull()]

,card,cost,count,tower_damage,damage,dps,death_damage,hit_speed,level,max_spawn,...,spawn_health,spawn_speed,spawner_health,troop_spawned,type,card_level,spawn_level,base_hp,shield_hp,can_attack_tower
57,Barbarian Hut,7.0,1.0,0.0,0.0,NaN,NaN,NaN,NaN,10.0,...,636.0,14.0,NaN,Barbarians,Spawners,7,9,NaN,NaN,0
58,Furnace,4.0,1.0,0.0,0.0,NaN,NaN,NaN,NaN,12.0,...,91.0,10.0,NaN,Fire Spirits,Spawners,7,9,NaN,NaN,0
59,Goblin Barrel,3.0,1.0,0.0,0.0,NaN,NaN,NaN,NaN,3.0,...,169.0,0.0,0.0,Goblins,Spawners,4,9,NaN,NaN,0
60,Goblin Hut,5.0,1.0,0.0,0.0,NaN,NaN,NaN,NaN,13.0,...,110.0,4.9,NaN,Spear Goblins,Spawners,7,9,NaN,NaN,0
61,Graveyard,5.0,1.0,0.0,0.0,NaN,NaN,NaN,NaN,15.0,...,67.0,0.5,0.0,Skeletons,Spawners,1,9,NaN,NaN,0
62,Tombstone,3.0,1.0,0.0,0.0,NaN,NaN,NaN,NaN,18.0,...,67.0,2.9,422.0,Skeletons,Spawners,7,9,NaN,NaN,0
63,Arrows,3.0,0.0,98.0,243.0,NaN,NaN,NaN,9.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1
64,Fireball,4.0,0.0,229.0,572.0,NaN,NaN,NaN,7.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1
65,Lightning,6.0,0.0,346.0,864.0,NaN,NaN,NaN,4.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1
66,Poison,4.0,0.0,240.0,600.0,NaN,NaN,NaN,4.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1


In [71]:
# here we cannot replace zero for spells as they deal high damage but at once so making it 0 will make loss of info 
df.loc[df['type'] != 'Damaging Spells', 'dps'] = df.loc[df['type'] != 'Damaging Spells', 'dps'].fillna(0)

In [74]:
df['death_damage'].unique()

array([  0., 272., 957., 259.,  53.,  70.,  nan])

In [76]:
# we fill this with 0s directly as it is specific feature 
df['death_damage'] = df['death_damage'].fillna(0)

In [80]:
df[df['hit_speed'].isnull()]
#Hit speed values were retained as null for cards where the attribute is not applicable, as assigning zero would misrepresent the attack mechanics.
# so we will keep hit speed as it is.

,card,cost,count,tower_damage,damage,dps,death_damage,hit_speed,level,max_spawn,...,spawn_health,spawn_speed,spawner_health,troop_spawned,type,card_level,spawn_level,base_hp,shield_hp,can_attack_tower
57,Barbarian Hut,7.0,1.0,0.0,0.0,0.0,0.0,NaN,NaN,10.0,...,636.0,14.0,NaN,Barbarians,Spawners,7,9,NaN,NaN,0
58,Furnace,4.0,1.0,0.0,0.0,0.0,0.0,NaN,NaN,12.0,...,91.0,10.0,NaN,Fire Spirits,Spawners,7,9,NaN,NaN,0
59,Goblin Barrel,3.0,1.0,0.0,0.0,0.0,0.0,NaN,NaN,3.0,...,169.0,0.0,0.0,Goblins,Spawners,4,9,NaN,NaN,0
60,Goblin Hut,5.0,1.0,0.0,0.0,0.0,0.0,NaN,NaN,13.0,...,110.0,4.9,NaN,Spear Goblins,Spawners,7,9,NaN,NaN,0
61,Graveyard,5.0,1.0,0.0,0.0,0.0,0.0,NaN,NaN,15.0,...,67.0,0.5,0.0,Skeletons,Spawners,1,9,NaN,NaN,0
62,Tombstone,3.0,1.0,0.0,0.0,0.0,0.0,NaN,NaN,18.0,...,67.0,2.9,422.0,Skeletons,Spawners,7,9,NaN,NaN,0
63,Arrows,3.0,0.0,98.0,243.0,NaN,0.0,NaN,9.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1
64,Fireball,4.0,0.0,229.0,572.0,NaN,0.0,NaN,7.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1
65,Lightning,6.0,0.0,346.0,864.0,NaN,0.0,NaN,4.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1
66,Poison,4.0,0.0,240.0,600.0,NaN,0.0,NaN,4.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1


In [84]:
# there shuld be level to each card so we fill it with median in place of missing values.
df['level'] = df['level'].fillna(df['level'].median())

In [87]:
df[df['radius'].notnull()]

,card,cost,count,tower_damage,damage,dps,death_damage,hit_speed,level,max_spawn,...,spawn_health,spawn_speed,spawner_health,troop_spawned,type,card_level,spawn_level,base_hp,shield_hp,can_attack_tower
63,Arrows,3.0,0.0,98.0,243.0,NaN,0.0,NaN,9.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1
64,Fireball,4.0,0.0,229.0,572.0,NaN,0.0,NaN,7.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1
65,Lightning,6.0,0.0,346.0,864.0,NaN,0.0,NaN,4.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1
66,Poison,4.0,0.0,240.0,600.0,NaN,0.0,NaN,4.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1
67,Rocket,6.0,0.0,493.0,1232.0,NaN,0.0,NaN,7.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1
68,The Log,2.0,0.0,96.0,240.0,NaN,0.0,NaN,1.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1
69,Zap,2.0,0.0,64.0,159.0,NaN,0.0,NaN,9.0,NaN,...,NaN,NaN,NaN,None,Damaging Spells,NaN,NaN,NaN,NaN,1


In [88]:
df['radius'] = df['radius'].fillna(0)

In [92]:
# 0 safely represents non-attacking cards
df['range'] = df['range'].fillna(0)

In [110]:
# only spawner cards have this feature so remaining are 0 
df['max_spawn'] = df['max_spawn'].fillna(0)

In [102]:
# filliing spawner stats with zero means cards do not have such ability 
spawn_cols = ['spawn_dps', 'spawn_damage', 'spawn_health', 'spawn_speed','spawner_health']

for col in spawn_cols:
    df[col] = df[col].fillna(0)

In [104]:
# simply level 0 does't exist in game so it means such feature is not applicable for these cards
df['card_level'] = df['card_level'].fillna(0)
df['spawn_level'] = df['spawn_level'].fillna(0)

In [106]:
# only some special cards have shield feature so remaining are filled with 0 
df['shield_hp'] = df['shield_hp'].fillna(0)

In [107]:
# we are keeping base hp as null as some cards like spells don't have health.

In [112]:
df.isnull().sum()

card                 0
cost                 0
count                0
tower_damage         0
damage               0
dps                  7
death_damage         0
hit_speed           13
level                0
max_spawn            0
radius               0
range                0
spawn_dps            0
spawn_damage         0
spawn_health         0
spawn_speed          0
spawner_health       0
troop_spawned        0
type                 0
card_level           0
spawn_level          0
base_hp             13
shield_hp            0
can_attack_tower     0
dtype: int64

In [114]:
# saving cleaned dataset for analysis
df.to_csv("cleaned_clash_data.csv", index=False)